libbeat/processors/add_id/generator/es_generator.go

```diff
@@ -0,0 +1,132 @@
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package generator
+
+import (
+	"encoding/base64"
+	"math/rand"
+	"sync"
+	"time"
+)
+
+type esTimeBasedUUIDGenerator struct{}
+
+// Singleton instance
+var _esTimeBasedUUIDGenerator IDGenerator = (*esTimeBasedUUIDGenerator)(nil)
+
+// ESTimeBasedUUIDGenerator returns the singleton instance for this generator
+func ESTimeBasedUUIDGenerator() IDGenerator {
+	return _esTimeBasedUUIDGenerator
+}
+
+var (
+	sequenceNumber uint32
+	lastTimestamp  uint64
+	delta          uint64
+	once           sync.Once
+	mac            []byte
+	mu             sync.Mutex
+)
+
+// NextID returns a base64-encoded, randomly-generated, but roughly ordered (over time), unique
+// ID. The algorithm used to generate the ID is the same as used by Elasticsearch.
+// See https://github.com/elastic/elasticsearch/blob/a666fb2266/server/src/main/java/org/elasticsearch/common/TimeBasedUUIDGenerator.java
+func (*esTimeBasedUUIDGenerator) NextID() string {
+	// Initialize sequence number and mac address. We do this here instead of doing it in a package-level
+	// init function to give the runtime time to generate enough entropy for randomization.
+	initOnce()
+
+	ts, seq := nextIDData()
+	var uuidBytes [15]byte
+
+	packID(uuidBytes[:], ts, seq)
+	return base64.RawURLEncoding.EncodeToString(uuidBytes[:])
+}
+
+func initOnce() {
+	once.Do(func() {
+		sequenceNumber = rand.Uint32()
+		m, err := getSecureMungedMACAddress()
+		if err != nil {
+			panic(err)
+		}
+		mac = m
+	})
+}
+
+func nextIDData() (uint64, uint32) {
+	mu.Lock()
+	defer mu.Unlock()
+
+	sequenceNumber++
+
+	// We only use bottom 3 bytes for the sequence number.
+	s := sequenceNumber & 0xffffff
+
+	now := uint64(time.Now().UnixNano() / 1000)
+	lastTimestamp = timestamp(now, lastTimestamp)
+	return lastTimestamp, s
+}
+
+// timestamp returns a monotonically-increasing timestamp (in ms) to use,
+// while accounting for system clock going backwards (e.g. due to a DST change).
+func timestamp(clockTS, lastTS uint64) uint64 {
+	// Last timestamp has not been previously initialized.
+	if lastTS == 0 {
+		return clockTS
+	}
+
+	// Normally clockTimestamp should be later than lastTimestamp. If that's the case, we can simply
+	// return clockTimestamp as the new timestamp.
+	if clockTS > lastTS {
+		return clockTS
+	}
+
+	// At this point, we know the system clock has gone backwards. So we increment the
+	// lastTimestamp by 1 (ms) and return it.
```

Comment:
microseconds, not milliseconds.

In order for lastTs to run faster than the clock we'd need 1M eps, which is quite unrealistic. But to be more on par with the ES implementation we maybe  should return `lastTs + 1` only if `sequence` has become 0.